src/program.rs

```diff
@@ -85,14 +86,45 @@ impl SBPFVersion {
 
 /// Holds the function symbols of an Executable
 #[derive(Debug, PartialEq, Eq)]
-pub struct FunctionRegistry<T> {
-    pub(crate) map: BTreeMap<u32, (Vec<u8>, T)>,
+pub enum FunctionRegistry<T> {
+    /// Sparse allows the registration of hashed symbols
+    Sparse(BTreeMap<u32, (Vec<u8>, T)>),
+    /// Dense allows the registration of symbols indexed by integers only
+    Dense(Vec<(Vec<u8>, T)>),
 }
 
-impl<T> Default for FunctionRegistry<T> {
-    fn default() -> Self {
-        Self {
-            map: BTreeMap::new(),
+impl<T> FunctionRegistry<T> {
```

Comment:
How about calling them `new_sparse` and `new_dense` and aliasing `default` to `new_sparse`? Effectively we only need `new_dense` in the new ELF loader and nowhere else.